aries_cloudagent/messaging/decorators/attach_decorator.py

```diff
@@ -24,15 +23,190 @@
+        description="signature",
+        required=True,
+        **BASE64URL_NO_PAD
+    )
+
+
+class AttachDecoratorDataJWS(BaseModel):
+    """
+    Detached JSON Web Signature for inclusion in attach decorator data.
+
+    May hold one signature in flattened format, or multiple signatures in the
+    "signatures" member.
+
+    """
+
+    class Meta:
+        """AttachDecoratorDataJWS metadata."""
+
+        schema_class = "AttachDecoratorDataJWSSchema"
+
+    def __init__(
+        self,
+        *,
+        header: AttachDecoratorDataJWSHeader = None,
+        protected: str = None,
+        signature: str = None,
+        signatures: Sequence[AttachDecoratorData1JWS] = None
+    ):
+        """Initialize JWS to include in attach decorator multi-sig data."""
+        self.header = header
+        self.protected = protected
+        self.signature = signature
+        self.signatures = signatures
+
+
+class AttachDecoratorDataJWSSchema(BaseModelSchema):
+    """Schema for detached JSON Web Signature for inclusion in attach decorator data."""
+
+    class Meta:
+        """Metadata for schema for detached JWS for inclusion in attach deco data."""
+
+        model_class = AttachDecoratorDataJWS
+
+    @pre_load
+    def validate_single_xor_multi_sig(self, data: Mapping, **kwargs):
+        """Ensure model is for either 1 or many sigatures, not mishmash of both."""
+
+        if "signatures" in data:
+            if any(k in data for k in ("header", "protected", "signature")):
+                raise BaseModelError(
+                    "AttachDecoratorDataJWSSchema: "
+                    "JWS must be flattened or general JSON serialization format"
+                )
+        elif not all(k in data for k in ("header", "signature")):
+            raise BaseModelError(
+                "AttachDecoratorDataJWSSchema: "
+                "Flattened JSON serialization format must include header and signature"
+            )
+
+        return data
+
+    header = fields.Nested(
+        AttachDecoratorDataJWSHeaderSchema,
+        required=False  # packed in signatures if multi-sig
+    )
+    protected = fields.Str(
+        description="protected JWS header",
+        required=False,  # packed in signatures if multi-sig
+        **BASE64URL_NO_PAD
+    )
+    signature = fields.Str(
+        description="signature",
+        required=False,  # packed in signatures if multi-sig
+        **BASE64URL_NO_PAD
+    )
+    signatures = fields.List(
+        fields.Nested(AttachDecoratorData1JWSSchema),
+        required=False,  # only present if multi-sig
+        description="List of signatures"
+    )
+
+
+def did_key(verkey: str) -> str:
+    """Qualify verkey into DID key if need be."""
+
+    return verkey if verkey.startswith("did:key:z") else f"did:key:z{verkey}"
+
+
+def raw_key(verkey: str) -> str:
+    """Strip qualified key to raw key if need be."""
+
+    return verkey[9:] if verkey.startswith("did:key:z") else verkey
```

Comment:
I don't think decoding a did:key is as simple as stripping off the prefix, I believe it adds a couple bytes to the binary representation and re-encodes as base58